satellite/admin/user.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (C) 2020 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package admin
+
+import (
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"golang.org/x/crypto/bcrypt"
+
+	"storj.io/common/uuid"
+	"storj.io/storj/satellite/console"
+)
+
+func (server *Server) addUser(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to read body: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	var input console.CreateUser
+
+	err = json.Unmarshal(body, &input)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to unmarshal request: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	if input.Email == "" {
+		http.Error(w, "Email is not set", http.StatusBadRequest)
+		return
+	}
+
+	if input.Password == "" {
+		http.Error(w, "Password is not set", http.StatusBadRequest)
+		return
+	}
+
+	hash, err := bcrypt.GenerateFromPassword([]byte(input.Password), 0)
+	if err != nil {
+		http.Error(w, "Unable to save password hash", http.StatusInternalServerError)
+		return
+	}
+
+	userID, err := uuid.New()
+	if err != nil {
+		http.Error(w, "Unable to create UUID", http.StatusInternalServerError)
+		return
+	}
+
+	user := console.CreateUser{
+		Email:    input.Email,
+		FullName: input.FullName,
+		Password: input.Password,
+	}
+
+	err = user.IsValid()
+	if err != nil {
+		http.Error(w, "User data is not valid", http.StatusBadRequest)
+		return
+	}
+
+	newuser, err := server.db.Console().Users().Insert(ctx, &console.User{
+		ID:           userID,
+		FullName:     user.FullName,
+		ShortName:    user.ShortName,
+		Email:        user.Email,
+		PasswordHash: hash,
+	})
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to insert user: %v", err), http.StatusInternalServerError)
+		return
+	}
+	//Set User Status to be activated, as we manually created it
+	newuser.Status = console.Active
+	newuser.PasswordHash = nil
```

Comment:
Why are we setting this to nil?

Reply:
We do not want to show it to the user/admin or override the password internally.

Reply:
We don't want to leak password data.